File: internal/stress/stress.go
// Package stress 实现生图压测:按自定义并发向中转站发起 N 次真实生图请求,
// 统计成功率/延迟/吞吐。压测只测量、不保存图片。
//
// Go 原生并发,直接用 goroutine 工作池实现,无 Python 版的 select 512 限制问题。
package stress

import (
	"context"
	"encoding/json"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/meihai66/stcs/internal/config"
	"github.com/meihai66/stcs/internal/generator"
)

type run struct {
	mu          sync.Mutex
	Status      string // running|done|cancelled|error
	Total       int
	Concurrency int
	Model       string
	Size        string
	Fmt         string
	Done        int
	OK          int
	Fail        int
	Latencies   []float64
	Errors      map[string]int
	StartMono   time.Time
	StartedAt   int64
	Elapsed     float64
	ErrMsg      string
	cancel      context.CancelFunc
}

var (
	gmu     sync.Mutex
	current *run
	httpcli = &http.Client{}
)

// State 返回当前压测的简单状态(供启动前判断是否在跑)。
func State() string {
	gmu.Lock()
	defer gmu.Unlock()
	if current == nil {
		return "idle"
	}
	current.mu.Lock()
	defer current.mu.Unlock()
	return current.Status
}

// StartParams 是一轮压测参数。
type StartParams struct {
	Prompt      string
	Total       int
	Concurrency int
	Model       string
	Size        string
	Quality     string
	Fmt         string
}

// Start 启动一轮压测(异步)。
func Start(sp StartParams) {
	cfg := config.Load()
	base, key := cfg.BaseURL, cfg.APIKey

	r := &run{
		Status: "running", Total: sp.Total, Concurrency: sp.Concurrency,
		Model: sp.Model, Size: sp.Size, Fmt: sp.Fmt,
		Errors: map[string]int{}, StartMono: time.Now(), StartedAt: time.Now().Unix(),
	}
	gmu.Lock()
	current = r
	gmu.Unlock()

	if base == "" || key == "" {
		r.Status = "error"
		r.ErrMsg = "未配置中转站地址或密钥"
		return
	}

	endpoint := strings.TrimRight(base, "/")
	if sp.Fmt == "chat" {
		endpoint += "/v1/chat/completions"
	} else {
		endpoint += "/v1/images/generations"
	}
	var payload map[string]any
	if sp.Fmt == "chat" {
		payload = map[string]any{
			"model":    sp.Model,
			"messages": []map[string]any{{"role": "user", "content": sp.Prompt}},
			"stream":   false,
		}
	} else {
		payload = map[string]any{"model": sp.Model, "prompt": sp.Prompt, "n": 1, "size": sp.Size}
		if sp.Quality != "" {
			payload["quality"] = sp.Quality
		}
	}
	body, _ := json.Marshal(payload)
	headers := map[string]string{"Authorization": "Bearer " + key, "Content-Type": "application/json"}
	timeout := time.Duration(cfg.Timeout) * time.Second

	conc := sp.Concurrency
	if conc > sp.Total {
		conc = sp.Total
	}
	r.Concurrency = conc

	ctx, cancel := context.WithCancel(context.Background())
	r.cancel = cancel

	go func() {
		var dispatched int64
		var dmu sync.Mutex
		var wg sync.WaitGroup
		for i := 0; i < conc; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				for {
					if ctx.Err() != nil {
						return
					}
					dmu.Lock()
					if dispatched >= int64(sp.Total) {
						dmu.Unlock()
						return
					}
					dispatched++
					dmu.Unlock()

					ok, lat, status, errLine := doOne(ctx, endpoint, body, headers, timeout, sp.Fmt)
					r.mu.Lock()
					r.Done++
					r.Latencies = append(r.Latencies, lat)
					if ok {
						r.OK++
					} else {
						r.Fail++
						label := errLine
						if status != 0 {
							label = "[" + itoa(status) + "] " + errLine
						}
						r.Errors[label]++
					}
					r.mu.Unlock()
				}
			}()
		}
		wg.Wait()
		r.mu.Lock()
		if r.Status == "running" {
			if ctx.Err() != nil {
				r.Status = "cancelled"
			} else {
				r.Status = "done"
			}
		}
		r.Elapsed = time.Since(r.StartMono).Seconds()
		r.mu.Unlock()
	}()
}

func doOne(ctx context.Context, endpoint string, body []byte, headers map[string]string, timeout time.Duration, fmtMode string) (ok bool, latMs float64, status int, errLine string) {
	start := time.Now()
	rctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	req, err := http.NewRequestWithContext(rctx, http.MethodPost, endpoint, strings.NewReader(string(body)))
	if err != nil {
		return false, ms(start), 0, "构造请求失败"
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := httpcli.Do(req)
	if err != nil {
		if ctx.Err() != nil {
			return false, ms(start), 0, "已取消"
		}
		if strings.Contains(err.Error(), "deadline") || strings.Contains(err.Error(), "timeout") {
			return false, ms(start), 0, "超时"
		}
		return false, ms(start), 0, "连接失败:" + err.Error()
	}
	defer resp.Body.Close()
	data := readAll(resp.Body)
	lat := ms(start)
	if resp.StatusCode != 200 {
		line := firstLine(generator.ExtractError(resp.StatusCode, data, resp.Header), 160)
		return false, lat, resp.StatusCode, line
	}
	if fmtMode != "chat" {
		var parsed struct {
			Data []json.RawMessage `json:"data"`
		}
		if json.Unmarshal(data, &parsed) != nil || len(parsed.Data) == 0 {
			return false, lat, 200, "200 但无图片数据"
		}
	}
	return true, lat, 200, ""
}

// Cancel 停止当前压测。
func Cancel() {
	gmu.Lock()
	r := current
	gmu.Unlock()
	if r == nil {
		return
	}
	r.mu.Lock()
	running := r.Status == "running"
	c := r.cancel
	r.mu.Unlock()
	if running && c != nil {
		c()
	}
}

// Stats 返回当前压测的统计数据。
func Stats() map[string]any {
	gmu.Lock()
	r := current
	gmu.Unlock()
	if r == nil {
		return map[string]any{"status": "idle"}
	}
	r.mu.Lock()
	defer r.mu.Unlock()

	lats := append([]float64(nil), r.Latencies...)
	sort.Float64s(lats)
	errs := map[string]int{}
	for k, v := range r.Errors {
		errs[k] = v
	}
	elapsed := r.Elapsed
	if elapsed == 0 && !r.StartMono.IsZero() {
		elapsed = time.Since(r.StartMono).Seconds()
	}
	rps := 0.0
	if elapsed > 0 {
		rps = round(float64(r.Done)/elapsed, 2)
	}
	out := map[string]any{
		"status":      r.Status,
		"total":       r.Total,
		"concurrency": r.Concurrency,
		"model":       r.Model,
		"size":        r.Size,
		"fmt":         r.Fmt,
		"done":        r.Done,
		"ok":          r.OK,
		"fail":        r.Fail,
		"elapsed":     round(elapsed, 2),
		"rps":         rps,
		"lat_min":     pickMin(lats),
		"lat_max":     pickMax(lats),
		"lat_avg":     avg(lats),
		"lat_p50":     pct(lats, 50),
		"lat_p95":     pct(lats, 95),
		"errors":      errs,
	}
	if r.ErrMsg != "" {
		out["error"] = r.ErrMsg
	}
	return out
}

// ---- helpers ----

func ms(start time.Time) float64 { return float64(time.Since(start).Microseconds()) / 1000.0 }

func readAll(r interface{ Read([]byte) (int, error) }) []byte {
	buf := make([]byte, 0, 1024)
	tmp := make([]byte, 4096)
	for {
		n, err := r.Read(tmp)
		buf = append(buf, tmp[:n]...)
		if err != nil {
			break
		}
		if len(buf) > 1<<20 { // 压测只看是否有数据,读 1MB 足够
			break
		}
	}
	return buf
}

func firstLine(s string, n int) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		s = s[:i]
	}
	if len(s) > n {
		s = s[:n]
	}
	return s
}

func pct(sorted []float64, p int) float64 {
	if len(sorted) == 0 {
		return 0
	}
	idx := p * len(sorted) / 100
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	if idx < 0 {
		idx = 0
	}
	return round(sorted[idx], 0)
}

func pickMin(s []float64) float64 {
	if len(s) == 0 {
		return 0
	}
	return round(s[0], 0)
}
func pickMax(s []float64) float64 {
	if len(s) == 0 {
		return 0
	}
	return round(s[len(s)-1], 0)
}
func avg(s []float64) float64 {
	if len(s) == 0 {
		return 0
	}
	var sum float64
	for _, v := range s {
		sum += v
	}
	return round(sum/float64(len(s)), 0)
}

func round(v float64, dec int) float64 {
	pow := 1.0
	for i := 0; i < dec; i++ {
		pow *= 10
	}
	return float64(int64(v*pow+0.5)) / pow
}

func itoa(i int) string {
	if i == 0 {
		return "0"
	}
	neg := i < 0
	if neg {
		i = -i
	}
	var b [20]byte
	pos := len(b)
	for i > 0 {
		pos--
		b[pos] = byte('0' + i%10)
		i /= 10
	}
	if neg {
		pos--
		b[pos] = '-'
	}
	return string(b[pos:])
}

File: GPT和Gemini生图带压测/stress.py
"""生图压测:按自定义并发向中转站发起 N 次真实生图请求,统计成功率/延迟/吞吐。

注意:压测会真实调用中转站、产生计费/消耗额度。为避免占满磁盘,压测只测量
不保存图片。结果保存在内存,前端轮询 /api/stress/status。

并发无硬上限:压测跑在**独立线程 + Proactor(IOCP)事件循环**里,绕开主服务
select 循环的 512 套接字限制,且与网页服务互不干扰(主服务照常用 select)。
"""
import asyncio
import sys
import threading
import time

import httpx

import config
import generator

_run: dict | None = None        # 当前压测状态(跨线程共享)
_loop = None                    # 压测线程的事件循环(供主线程取消任务用)
_thread = None                  # 压测线程
_lock = threading.Lock()        # 保护 _run 里计数/集合的读写


def state() -> dict | None:
    return _run


async def _one(client, endpoint, payload, headers, timeout, fmt):
    """发一次请求,返回 (ok, latency_ms, status_code, error)。压测不重试,要看真实结果。"""
    start = time.monotonic()
    try:
        resp = await client.post(endpoint, json=payload, headers=headers, timeout=timeout)
        lat = (time.monotonic() - start) * 1000
        if resp.status_code != 200:
            line = generator._extract_error(resp).splitlines()[0][:160]
            return False, lat, resp.status_code, line
        ok = True
        try:
            body = resp.json()
            if fmt != "chat":
                ok = bool(body.get("data"))
        except ValueError:
            ok = fmt == "chat"
        return ok, lat, 200, (None if ok else "200 但无图片数据")
    except httpx.TimeoutException:
        return False, (time.monotonic() - start) * 1000, 0, "超时"
    except httpx.RequestError as e:
        return False, (time.monotonic() - start) * 1000, 0, f"连接失败:{e}"


async def _run_all(endpoint, payload, headers, timeout, total, concurrency, fmt):
    """工作池:只开 concurrency 个 worker,各自循环领取任务,真正限制同时在飞数。"""
    limits = httpx.Limits(max_connections=concurrency, max_keepalive_connections=concurrency)
    dispatched = 0  # 同一事件循环线程内,「判断+自增」之间无 await,天然原子

    async with httpx.AsyncClient(limits=limits) as client:
        async def worker():
            nonlocal dispatched
            while not _run["cancel"]:
                if dispatched >= total:
                    return
                dispatched += 1
                ok, lat, status, err = await _one(client, endpoint, payload, headers, timeout, fmt)
                with _lock:
                    _run["done"] += 1
                    _run["latencies"].append(lat)
                    if ok:
                        _run["ok"] += 1
                    else:
                        _run["fail"] += 1
                        label = f"[{status}] {err}" if status else str(err)
                        _run["errors"][label] = _run["errors"].get(label, 0) + 1

        tasks = [asyncio.create_task(worker()) for _ in range(concurrency)]
        _run["_tasks"] = tasks
        await asyncio.gather(*tasks, return_exceptions=True)


def start(*, prompt, total, concurrency, model, size, quality, fmt):
    """启动一轮压测:在独立线程里用 Proactor 循环跑,绕过主服务 select 的 512 限制。"""
    global _run, _loop, _thread
    cfg = config.load()
    base, key = cfg.get("base_url"), cfg.get("api_key")
    if not base or not key:
        _run = {"status": "error", "error": "未配置中转站地址或密钥", "done": 0, "total": total}
        return

    endpoint = base.rstrip("/") + ("/v1/chat/completions" if fmt == "chat" else "/v1/images/generations")
    headers = {"Authorization": f"Bearer {key}", "Content-Type": "application/json"}
    if fmt == "chat":
        payload = {"model": model, "messages": [{"role": "user", "content": prompt}], "stream": False}
    else:
        payload = {"model": model, "prompt": prompt, "n": 1, "size": size}
        if quality:
            payload["quality"] = quality
    timeout = int(cfg.get("timeout", 300))
    concurrency = min(concurrency, total)  # 并发不必超过总数

    _run = {
        "status": "running", "total": total, "concurrency": concurrency,
        "model": model, "size": size, "fmt": fmt,
        "done": 0, "ok": 0, "fail": 0, "latencies": [], "errors": {},
        "start_mono": time.monotonic(), "started_at": int(time.time()),
        "cancel": False, "_tasks": [],
    }

    def thread_main():
        global _loop
        loop = asyncio.ProactorEventLoop() if sys.platform == "win32" else asyncio.new_event_loop()
        _loop = loop
        asyncio.set_event_loop(loop)
        try:
            loop.run_until_complete(_run_all(endpoint, payload, headers, timeout, total, concurrency, fmt))
        except Exception as e:  # 兜底,别让线程静默死掉
            _run["status"] = "error"
            _run["error"] = f"压测线程异常:{e}"
        finally:
            try:
                loop.close()
            except Exception:
                pass
            _loop = None
            if _run.get("status") == "running":
                _run["status"] = "cancelled" if _run.get("cancel") else "done"
            _run["elapsed"] = time.monotonic() - _run["start_mono"]
            _run.pop("_tasks", None)

    _thread = threading.Thread(target=thread_main, daemon=True)
    _thread.start()


def cancel():
    """停止:置取消标志。worker 跑完当前这次请求后就退出,连接干净关闭。

    不用跨线程 task.cancel():在 Proactor 循环上强行取消在飞的 httpx 请求会让
    连接池清理卡死。靠标志位让 worker 自然收尾,最多多等一次请求的时间。
    """
    if _run and _run.get("status") == "running":
        _run["cancel"] = True


def _pct(sorted_lats, p):
    if not sorted_lats:
        return 0
    idx = min(len(sorted_lats) - 1, max(0, int(p / 100 * len(sorted_lats))))
    return sorted_lats[idx]


def stats() -> dict:
    """计算当前压测的统计数据(供 /api/stress/status 返回)。对任何 _run 形态都安全。"""
    if not _run:
        return {"status": "idle"}
    # 压测线程在另一个线程写 _run,这里加锁快照,避免「list/dict changed during iteration」
    with _lock:
        lats = sorted(_run.get("latencies", []))
        errors = dict(_run.get("errors", {}))
        done = _run.get("done", 0)
        ok = _run.get("ok", 0)
        fail = _run.get("fail", 0)
    start_mono = _run.get("start_mono")
    elapsed = _run.get("elapsed") or ((time.monotonic() - start_mono) if start_mono else 0)
    return {
        "status": _run.get("status", "idle"),
        "total": _run.get("total", 0),
        "concurrency": _run.get("concurrency", 0),
        "model": _run.get("model", ""), "size": _run.get("size", ""), "fmt": _run.get("fmt", ""),
        "done": done, "ok": ok, "fail": fail,
        "elapsed": round(elapsed, 2),
        "rps": round(done / elapsed, 2) if elapsed > 0 else 0,
        "lat_min": round(min(lats), 0) if lats else 0,
        "lat_max": round(max(lats), 0) if lats else 0,
        "lat_avg": round(sum(lats) / len(lats), 0) if lats else 0,
        "lat_p50": round(_pct(lats, 50), 0),
        "lat_p95": round(_pct(lats, 95), 0),
        "errors": errors,
        "error": _run.get("error"),
    }
